### PR TITLE
[#3980]feat(flink-connector): Support partition for hive table operation

### PR DESCRIPTION
**Pull request by coolderli**
add UTs

### What changes were proposed in this pull request?

- support partitions for creating the hive table

### Why are the changes needed?

- Fix: #3980 

### Does this PR introduce _any_ user-facing change?

- no

### How was this patch tested?

- add ITs

### CODE REVIEW

**Comment by coolderli** [comment]
@FANNG1 Please help review this. Thanks.

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java [hunk 1]
```diff
@@ -77,7 +78,7 @@
  * The BaseCatalog that provides a default implementation for all methods in the {@link
  * org.apache.flink.table.catalog.Catalog} interface.
  */
-public abstract class BaseCatalog extends AbstractCatalog {
+public abstract class BaseCatalog extends AbstractCatalog implements TransformConverter {
```
Comment (FANNG1) [inline flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java:81]: better to use a specific `transformConverter` like `propertiesConverter`?

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/TransformConverter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            .filter(t -> t instanceof Transforms.IdentityTransform)
+            .map(Transform::name)
+            .collect(Collectors.toList());
+    Preconditions.checkArgument(
+        partitionKeys.size() == transforms.length,
+        "Flink only support identity transform for now.");
+    return partitionKeys;
+  }
+
+  default Transform[] toGravitinoTransforms(List<String> partitionsKey) {
```
Comment (FANNG1) [inline flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/TransformConverter.java:47]: `toGravitinoPartitions` to make it more clear?

File: flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java [hunk 3]
```diff
@@ -281,6 +291,59 @@ public void testGetCatalogFromGravitino() {
...
+                Column.of("double_type", Types.DoubleType.get(), "double_type")
+              };
+          assertColumns(columns, table.columns());
+          Transform[] partitions =
+              new Transform[] {
+                Transforms.identity("string_type"), Transforms.identity("double_type")
+              };
+          Assertions.assertArrayEquals(partitions, table.partitioning());
+
+          TestUtils.assertTableResult(
```
Comment (FANNG1) [inline flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java:334]: could you check the partition dir exists?

File: flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java [hunk 4]
```diff
@@ -281,6 +291,59 @@ public void testGetCatalogFromGravitino() {
...
+                  "CREATE TABLE %s "
+                      + "(string_type STRING COMMENT 'string_type', "
+                      + " double_type DOUBLE COMMENT 'double_type')"
+                      + " PARTITIONED BY (string_type, double_type)"
+                      + " COMMENT '%s' WITH ("
+                      + "'%s' = '%s')",
+                  tableName, comment, key, value);
+          TestUtils.assertTableResult(result, ResultKind.SUCCESS);
+
+          Table table =
```
Comment (FANNG1) [inline flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java:317]: The test covers `create hive partition table path`, but not covers the `load hive partition table path` in flink connector, YES?

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java [hunk 5]
```diff
@@ -79,10 +80,12 @@
  */
 public abstract class BaseCatalog extends AbstractCatalog {
   private final PropertiesConverter propertiesConverter;
+  private final TransformConverter transformConverter;
```
Comment (FANNG1) [inline flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java:83]: does it will transform distribution , sortorders? if not,  `PartitionConverter` may better

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/DefaultTransformConverter.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
```
Comment (FANNG1) [inline flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/DefaultTransformConverter.java:2]: please use apache licence header

**Comment by FANNG1** [comment]
LGTM, except few comments

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java (hunk 1 above)
Comment (coolderli) [reply flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java]: Done

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/TransformConverter.java (hunk 2 above)
Comment (coolderli) [reply flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/TransformConverter.java]: Done

File: flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java (hunk 3 above)
Comment (coolderli) [reply flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java:355]: Done

File: flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java (hunk 4 above)
Comment (coolderli) [reply flink-connector/src/test/java/com/datastrato/gravitino/flink/connector/integration/test/hive/FlinkHiveCatalogIT.java:323]: Done

**Comment by coolderli** [comment]
@FANNG1 I fixed it. Thanks.

**Comment by FANNG1** [comment]
@coolderli , seems you add some partittion interfaces support since last review  like `listPartitions`, is it expected?

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java [hunk 7]
```diff
@@ -280,37 +287,36 @@ public void alterTable(ObjectPath objectPath, CatalogBaseTable catalogBaseTable,
   }
 
   @Override
-  public List<CatalogPartitionSpec> listPartitions(ObjectPath objectPath)
+  public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)
       throws TableNotExistException, TableNotPartitionedException, CatalogException {
-    throw new UnsupportedOperationException();
```
Comment (FANNG1) [inline flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java:285]: Could you use the partition interface in Gravitino?

**Comment by coolderli** [comment]
> @coolderli , seems you add some partittion interfaces support since last review like `listPartitions`, is it expected?

@FANNG1 I think it's necessary. I meet an exception as follows. So I have to implement it.
```
FlinkHiveCatalogIT > testHivePartitionTable() FAILED
    java.lang.UnsupportedOperationException
        at org.apache.gravitino.flink.connector.catalog.BaseCatalog.listPartitions(BaseCatalog.java:292)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.getPartitionsTableStats(FlinkRecomputeStatisticsProgram.java:180)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.recomputeStatistics(FlinkRecomputeStatisticsProgram.java:144)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.recomputeStatistics(FlinkRecomputeStatisticsProgram.java:100)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.access$000(FlinkRecomputeStatisticsProgram.java:66)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram$1.visit(FlinkRecomputeStatisticsProgram.java:75)
        at org.apache.calcite.rel.core.TableScan.accept(TableScan.java:180)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.$anonfun$visit$1(RelShuttles.scala:37)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
        at scala.collection.Iterator.foreach(Iterator.scala:937)
        at scala.collection.Iterator.foreach$(Iterator.scala:937)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
        at scala.collection.IterableLike.foreach(IterableLike.scala:70)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike.map(TraversableLike.scala:233)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.visit(RelShuttles.scala:35)
        at org.apache.calcite.rel.RelHomogeneousShuttle.visit(RelHomogeneousShuttle.java:66)
        at org.apache.calcite.rel.logical.LogicalProject.accept(LogicalProject.java:135)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.$anonfun$visit$1(RelShuttles.scala:37)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
        at scala.collection.Iterator.foreach(Iterator.scala:937)
        at scala.collection.Iterator.foreach$(Iterator.scala:937)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
        at scala.collection.IterableLike.foreach(IterableLike.scala:70)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike.map(TraversableLike.scala:233)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.visit(RelShuttles.scala:35)
        at org.apache.calcite.rel.RelHomogeneousShuttle.visit(RelHomogeneousShuttle.java:90)
        at org.apache.calcite.rel.logical.LogicalSort.accept(LogicalSort.java:86)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.$anonfun$visit$1(RelShuttles.scala:37)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
        at scala.collection.Iterator.foreach(Iterator.scala:937)
        at scala.collection.Iterator.foreach$(Iterator.scala:937)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
        at scala.collection.IterableLike.foreach(IterableLike.scala:70)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike.map(TraversableLike.scala:233)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.utils.DefaultRelShuttle.visit(RelShuttles.scala:35)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.optimize(FlinkRecomputeStatisticsProgram.java:80)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkRecomputeStatisticsProgram.optimize(FlinkRecomputeStatisticsProgram.java:66)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkGroupProgram.$anonfun$optimize$2(FlinkGroupProgram.scala:59)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:156)
        at scala.collection.Iterator.foreach(Iterator.scala:937)
        at scala.collection.Iterator.foreach$(Iterator.scala:937)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
        at scala.collection.IterableLike.foreach(IterableLike.scala:70)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:154)
        at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkGroupProgram.$anonfun$optimize$1(FlinkGroupProgram.scala:56)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkGroupProgram.$anonfun$optimize$1$adapted(FlinkGroupProgram.scala:51)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:156)
        at scala.collection.immutable.Range.foreach(Range.scala:155)
        at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:154)
        at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkGroupProgram.optimize(FlinkGroupProgram.scala:51)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.$anonfun$optimize$1(FlinkChainedProgram.scala:59)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:156)
        at scala.collection.Iterator.foreach(Iterator.scala:937)
        at scala.collection.Iterator.foreach$(Iterator.scala:937)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
        at scala.collection.IterableLike.foreach(IterableLike.scala:70)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:156)
        at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:154)
        at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
        at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.optimize(FlinkChainedProgram.scala:55)
        at org.apache.flink.table.planner.plan.optimize.BatchCommonSubGraphBasedOptimizer.optimizeTree(BatchCommonSubGraphBasedOptimizer.scala:93)
        at org.apache.flink.table.planner.plan.optimize.BatchCommonSubGraphBasedOptimizer.optimizeBlock(BatchCommonSubGraphBasedOptimizer.scala:58)
        at org.apache.flink.table.planner.plan.optimize.BatchCommonSubGraphBasedOptimizer.$anonfun$doOptimize$1(BatchCommonSubGraphBasedOptimizer.scala:45)
        at org.apache.flink.table.planner.plan.optimize.BatchCommonSubGraphBasedOptimizer.$anonfun$doOptimize$1$adapted(BatchCommonSubGraphBasedOptimizer.scala:45)
        at scala.collection.immutable.List.foreach(List.scala:388)
        at org.apache.flink.table.planner.plan.optimize.BatchCommonSubGraphBasedOptimizer.doOptimize(BatchCommonSubGraphBasedOptimizer.scala:45)
        at org.apache.flink.table.planner.plan.optimize.CommonSubGraphBasedOptimizer.optimize(CommonSubGraphBasedOptimizer.scala:87)
        at org.apache.flink.table.planner.delegation.PlannerBase.optimize(PlannerBase.scala:324)
        at org.apache.flink.table.planner.delegation.PlannerBase.translate(PlannerBase.scala:182)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.translate(TableEnvironmentImpl.java:1277)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeQueryOperation(TableEnvironmentImpl.java:1055)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1120)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:735)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.sql(FlinkEnvIT.java:159)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.lambda$testHivePartitionTable$3(FlinkHiveCatalogIT.java:360)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.doWithSchema(FlinkEnvIT.java:175)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.testHivePartitionTable(FlinkHiveCatalogIT.java:307)


```

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java (hunk 7 above)
Comment (coolderli) [reply flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java]: @FANNG1 I'd like to create an issue to fix it and add more integration tests. What do you think?

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+import org.apache.gravitino.rel.expressions.transforms.Transform;
+
+/**
+ * The PartitionConverter is used to convert the partition between Flink and Gravitino. The Flink
+ * only support identity transform. Some of the table like Apache Paimon will use the table
+ * properties to store the partition transform, so we can implement this interface to achieve more
+ * partition transform.
+ */
+public abstract class PartitionConverter {
```
Comment (FANNG1) [inline flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java:30]: interface?

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java [hunk 9]
```diff
@@ -347,9 +353,9 @@ public List<String> listFunctions(String s) throws DatabaseNotExistException, Ca
   }
 
   @Override
-  public CatalogFunction getFunction(ObjectPath objectPath)
+  public CatalogFunction getFunction(ObjectPath tablePath)
       throws FunctionNotExistException, CatalogException {
-    throw new UnsupportedOperationException();
```
Comment (FANNG1) [inline flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java:352]: `getFunction` `getXXStatistics` seems not related to this PR?

File: flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java (hunk 7 above)
Comment (FANNG1) [reply flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java]: could you create an issue?

Comment (coolderli) [reply flink-connector/src/main/java/com/datastrato/gravitino/flink/connector/catalog/BaseCatalog.java]: https://github.com/apache/gravitino/issues/4216

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java (hunk 8 above)
Comment (coolderli) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java:30]: @FANNG1 I think the abstract class is OK. Other classes won't implement it. What do you think?

Comment (FANNG1) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java:30]: interface is more proper here to me

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java (hunk 9 above)
Comment (coolderli) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java:352]: getTableStatistics is needed.
```
FlinkHiveCatalogIT > testHivePartitionTable() FAILED
    org.apache.flink.table.api.ValidationException: SQL validation failed. null
        at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.org$apache$flink$table$planner$calcite$FlinkPlannerImpl$$validate(FlinkPlannerImpl.scala:200)
        at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.validate(FlinkPlannerImpl.scala:117)
        at org.apache.flink.table.planner.operations.SqlNodeToOperationConversion.convert(SqlNodeToOperationConversion.java:261)
        at org.apache.flink.table.planner.delegation.ParserImpl.parse(ParserImpl.java:106)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:728)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.sql(FlinkEnvIT.java:159)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.lambda$testHivePartitionTable$3(FlinkHiveCatalogIT.java:360)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.doWithSchema(FlinkEnvIT.java:175)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.testHivePartitionTable(FlinkHiveCatalogIT.java:307)

        Caused by:
        java.lang.UnsupportedOperationException
            at org.apache.gravitino.flink.connector.catalog.BaseCatalog.getTableStatistics(BaseCatalog.java:387)
            at org.apache.flink.table.planner.catalog.DatabaseCalciteSchema.extractTableStats(DatabaseCalciteSchema.java:122)
            at org.apache.flink.table.planner.catalog.DatabaseCalciteSchema.getStatistic(DatabaseCalciteSchema.java:106)
            at org.apache.flink.table.planner.catalog.DatabaseCalciteSchema.lambda$getTable$0(DatabaseCalciteSchema.java:94)
            at java.util.Optional.map(Optional.java:215)
            at org.apache.flink.table.planner.catalog.DatabaseCalciteSchema.getTable(DatabaseCalciteSchema.java:90)
            at org.apache.calcite.jdbc.SimpleCalciteSchema.getImplicitTable(SimpleCalciteSchema.java:126)
            at org.apache.calcite.jdbc.CalciteSchema.getTable(CalciteSchema.java:295)
            at org.apache.calcite.sql.validate.EmptyScope.resolve_(EmptyScope.java:145)
            at org.apache.calcite.sql.validate.EmptyScope.resolveTable(EmptyScope.java:101)
            at org.apache.calcite.sql.validate.DelegatingScope.resolveTable(DelegatingScope.java:214)
            at org.apache.calcite.sql.validate.IdentifierNamespace.resolveImpl(IdentifierNamespace.java:116)
            at org.apache.calcite.sql.validate.IdentifierNamespace.validateImpl(IdentifierNamespace.java:188)
            at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1050)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1025)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateFrom(SqlValidatorImpl.java:3294)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateFrom(SqlValidatorImpl.java:3276)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3603)
            at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:64)
            at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1050)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1025)
            at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:248)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1000)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:749)
            at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.org$apache$flink$table$planner$calcite$FlinkPlannerImpl$$validate(FlinkPlannerImpl.scala:196)
            ... 8 more

```

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java (hunk 8 above)
Comment (coolderli) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/PartitionConverter.java]: Done

**Comment by coolderli** [comment]
@FANNG1 I removed the `getFunction` and kept `getXXStatistic`. Please take a look. Thanks.

**Comment by coolderli** [comment]
`getFunction` is also needed.
```
FlinkHiveCatalogIT > testHivePartitionTable() FAILED
    org.apache.flink.table.api.ValidationException: SQL validation failed. null
        at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.org$apache$flink$table$planner$calcite$FlinkPlannerImpl$$validate(FlinkPlannerImpl.scala:200)
        at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.validate(FlinkPlannerImpl.scala:117)
        at org.apache.flink.table.planner.operations.SqlNodeToOperationConversion.convert(SqlNodeToOperationConversion.java:261)
        at org.apache.flink.table.planner.delegation.ParserImpl.parse(ParserImpl.java:106)
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:728)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.sql(FlinkEnvIT.java:159)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.lambda$testHivePartitionTable$3(FlinkHiveCatalogIT.java:360)
        at org.apache.gravitino.flink.connector.integration.test.FlinkEnvIT.doWithSchema(FlinkEnvIT.java:175)
        at org.apache.gravitino.flink.connector.integration.test.hive.FlinkHiveCatalogIT.testHivePartitionTable(FlinkHiveCatalogIT.java:307)

        Caused by:
        java.lang.UnsupportedOperationException
            at org.apache.gravitino.flink.connector.catalog.BaseCatalog.getFunction(BaseCatalog.java:358)
            at org.apache.flink.table.catalog.FunctionCatalog.resolvePreciseFunctionReference(FunctionCatalog.java:617)
            at org.apache.flink.table.catalog.FunctionCatalog.resolvePreciseFunctionReference(FunctionCatalog.java:593)
            at org.apache.flink.table.catalog.FunctionCatalog.lambda$resolveAmbiguousFunctionReference$5(FunctionCatalog.java:671)
            at java.util.Optional.orElseGet(Optional.java:267)
            at org.apache.flink.table.catalog.FunctionCatalog.resolveAmbiguousFunctionReference(FunctionCatalog.java:671)
            at org.apache.flink.table.catalog.FunctionCatalog.lookupFunction(FunctionCatalog.java:424)
            at org.apache.flink.table.planner.catalog.FunctionCatalogOperatorTable.lookupOperatorOverloads(FunctionCatalogOperatorTable.java:99)
            at org.apache.calcite.sql.util.ChainedSqlOperatorTable.lookupOperatorOverloads(ChainedSqlOperatorTable.java:69)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.makeNullaryCall(SqlValidatorImpl.java:1829)
            at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.visit(SqlValidatorImpl.java:6430)
            at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.visit(SqlValidatorImpl.java:6414)
            at org.apache.calcite.sql.SqlIdentifier.accept(SqlIdentifier.java:324)
            at org.apache.calcite.sql.validate.SqlValidatorImpl$Expander.go(SqlValidatorImpl.java:6423)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.expand(SqlValidatorImpl.java:6004)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateOrderList(SqlValidatorImpl.java:4150)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3621)
            at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:64)
            at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1050)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1025)
            at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:248)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1000)
            at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:749)
            at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.org$apache$flink$table$planner$calcite$FlinkPlannerImpl$$validate(FlinkPlannerImpl.scala:196)

```

File: flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java (hunk 9 above)
Comment (FANNG1) [reply flink-connector/src/main/java/org/apache/gravitino/flink/connector/catalog/BaseCatalog.java:352]: got it

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@coolderli , merged to main, thanks for your contribution!
